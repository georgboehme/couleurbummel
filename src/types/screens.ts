import {DrawerScreenProps} from '@react-navigation/drawer';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import React from 'react';
import {Region} from 'react-native-maps/lib/sharedTypes';

import {Address, City, Corporation} from './model';

export type DrawerScreenParamsList = {
  Map: {
    region?: Region;
    addressIds?: string[];
  };
  ExploreNearby: undefined;
  ExploreCities: undefined;
  ExploreCountries: undefined;
  ExploreOrganisations: undefined;
  SearchName: undefined;
  SearchColours: undefined;
  Settings: undefined;
};

export type DrawerScreenName = keyof DrawerScreenParamsList;

export type ScreenProps = DrawerScreenProps<
  DrawerScreenParamsList,
  DrawerScreenName
>;

export type MapScreenProps = DrawerScreenProps<DrawerScreenParamsList, 'Map'>;
export type ColoursSearchScreenProps = DrawerScreenProps<
  DrawerScreenParamsList,
  'SearchColours'
>;
export type SettingsScreenProps = DrawerScreenProps<
  DrawerScreenParamsList,
  'Settings'
>;

export type CorporationListType = 'searchable' | 'non-searchable' | 'fragment';

export interface CorporationListProps {
  addressesAndCorporations: [Address, Corporation][];
  parentScreenName: ScreenWithCorporationList;
  searchable: boolean;
  searchOptions?: {
    showsResultsOnEmptySearch?: boolean;
  };
  headerTitle?: string;
  hideOrganisation?: boolean;
  headerComponent?: React.ReactElement<{}>;
  testID?: string;
}

export interface CityListProps {
  cities?: City[];
  testID?: string;
}

export type StackScreenParamsList = {
  CorporationList: CorporationListProps;
  Search: undefined;
  Cities: CityListProps;
  CorporationDetails: {corporation: Corporation};
  Countries: undefined;
  Organisations: undefined;
};

export type ExploreScreen =
  | 'Cities'
  | 'Countries'
  | 'Organisations'
  | 'Closest';

export type ScreenWithCorporationList =
  | ExploreScreen
  | 'SearchName'
  | 'SearchColours';

export type SearchScreenProps = NativeStackScreenProps<
  StackScreenParamsList,
  'Search'
>;

export type CorporationListScreenProps = NativeStackScreenProps<
  StackScreenParamsList,
  'CorporationList'
>;

export type CorporationDetailsScreenProps = NativeStackScreenProps<
  StackScreenParamsList,
  'CorporationDetails'
>;

export type CitiesScreenProps = NativeStackScreenProps<
  StackScreenParamsList,
  'Cities'
>;

export type CountriesProps = NativeStackScreenProps<
  StackScreenParamsList,
  'Countries'
>;

export type OrganisationProps = NativeStackScreenProps<
  StackScreenParamsList,
  'Organisations'
>;
