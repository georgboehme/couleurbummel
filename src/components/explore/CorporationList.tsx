import {StackActions, useNavigation} from '@react-navigation/native';
import {ListItem, makeStyles} from '@rneui/themed';
import React, {useContext, useEffect, useState} from 'react';
import {useTranslation} from 'react-i18next';
import {FlatList} from 'react-native';
import {LatLng} from 'react-native-maps/lib/sharedTypes';

import constants from '../../constants';
import log from '../../log';
import {Address, Corporation, Organisation} from '../../types/model';
import {
  CorporationListProps,
  CorporationListScreenProps,
} from '../../types/screens';
import {compareDistances, formatDistance} from '../../util';
import {ColoursSet} from '../ColoursSet';
import {toggleFavouriteCorporationAction} from '../contexts/actions';
import {StateContext} from '../contexts/GlobalStateContext';
import {ModelContext} from '../contexts/ModelContext';
import {UserLocationContext} from '../contexts/UserLocationContext';
import {FavouriteButton} from '../FavouriteButton';
import fuzzySearch from '../fuzzySearch';
import {SearchableList} from '../SearchableList';

const sortCorporations = (
  corporations: [Address, Corporation][],
  option: 'distance' | 'name',
  userLocation?: LatLng,
) => {
  const copy = [...corporations];

  copy.sort(([address1, corp1], [address2, corp2]) => {
    if (option === 'distance' && userLocation) {
      return compareDistances(address1, address2, userLocation);
    }

    return corp1.shortName.localeCompare(corp2.shortName);
  });

  return copy;
};

const filterCorporations = (
  searchTerm: string,
  corporations: [Address, Corporation][],
) =>
  corporations.filter(([_address, c]) =>
    fuzzySearch(searchTerm, c.shortName.toLowerCase()),
  );

const renderCorporation =
  (hideOrganisation?: boolean) => (item: [Address, Corporation]) => {
    const [address, corporation] = item;

    return (
      <CorporationItem
        corporation={corporation}
        address={address}
        hideOrganisation={hideOrganisation}
      />
    );
  };

/**
 * A list of corporations
 */
export const CorporationList = (props: CorporationListProps) => {
  const styles = useStyles();
  const {globalState} = useContext(StateContext);
  const userLocation = useContext(UserLocationContext);

  const navigation = useNavigation();

  const [addressesAndCorporations, setAddressesAndCorporations] = useState(
    props.addressesAndCorporations,
  );

  useEffect(() => {
    if (props.headerTitle) {
      navigation.setOptions({title: props.headerTitle});
    }
  });

  useEffect(() => {
    log.debug('State changed: CorporationList');
    const option = globalState.corporationSorting[props.parentScreenName];

    log.debug(`Sorting ${props.parentScreenName} by ${option}`);

    setAddressesAndCorporations(
      sortCorporations(props.addressesAndCorporations, option, userLocation),
    );
  }, [globalState, userLocation, props]);

  if (!props.searchable) {
    return (
      <FlatList
        style={styles.container}
        ListHeaderComponent={props.headerComponent}
        testID={props.testID}
        data={addressesAndCorporations}
        renderItem={({item}) => renderCorporation(props.hideOrganisation)(item)}
        keyExtractor={([_address, c]) => c.id}
      />
    );
  }

  return (
    <SearchableList
      testID={props.testID}
      allItems={addressesAndCorporations}
      filterItems={filterCorporations}
      renderItem={renderCorporation(props.hideOrganisation)}
      keyExtractor={([_address, c]) => c.id}
      searchDelayMs={constants.explore.search.delayMs}
      headerComponent={props.headerComponent}
      showsAllResultsOnEmptySearch={
        props.searchOptions?.showsResultsOnEmptySearch
      }
    />
  );
};

/**
 * A row in a list of corporations
 *
 * @param corporation the corporation
 * @param address the address at which the corporation is located
 * @param hideOrganisation whether we want to hide the corporation's organisation name in parentheses
 * @constructor
 */
const CorporationItem = ({
  corporation,
  address,
  hideOrganisation,
}: {
  corporation: Corporation;
  address: Address;
  hideOrganisation?: boolean;
}) => {
  const navigation = useNavigation();
  const styles = useStyles();
  const {t} = useTranslation();

  const model = useContext(ModelContext);
  const userLocation = useContext(UserLocationContext);

  let orgsString = '';
  if (!hideOrganisation) {
    const orgs = corporation.organisationIds
      ?.map(orgId => model.organisation(orgId))
      .filter(o => !!o)
      .map(o => (o as Organisation).abbreviation);

    orgsString = orgs ? ` (${orgs.join(', ')})` : '';
  }

  return (
    <ListItem.Swipeable
      key={corporation.id}
      onPress={() => {
        log.debug(corporation.shortName);
        navigation.dispatch(
          StackActions.push('CorporationDetails', {
            corporation: corporation,
          }),
        );
      }}
      style={({pressed}) =>
        pressed ? styles.listItemPressed : styles.listItemDefault
      }
      rightContent={reset => (
        <FavouriteButton
          itemId={corporation.id}
          toggleAction={toggleFavouriteCorporationAction}
          postDispatch={reset}
        />
      )}
      bottomDivider>
      <ListItem.Content>
        <ListItem.Title style={styles.corporationName}>
          {`${corporation.shortName}${orgsString}`}
        </ListItem.Title>
        {userLocation && (
          <ListItem.Subtitle style={styles.distanceText}>
            {t('DISTANCE', {
              distanceString: formatDistance(userLocation, {
                latitude: address.latitude,
                longitude: address.longitude,
              }),
            })}
          </ListItem.Subtitle>
        )}
        <ColoursSet c={corporation} />
      </ListItem.Content>
      <ListItem.Chevron />
    </ListItem.Swipeable>
  );
};

export const CorporationListScreen = ({route}: CorporationListScreenProps) => {
  return (
    <CorporationList
      addressesAndCorporations={route.params.addressesAndCorporations}
      parentScreenName={route.params.parentScreenName}
      searchable={route.params.searchable}
      searchOptions={route.params.searchOptions}
    />
  );
};

const useStyles = makeStyles(theme => ({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  corporationName: {},
  distanceText: {
    color: theme.colors.grey0,
    fontSize: 11,
    fontStyle: 'italic',
  },
  listItemPressed: {
    backgroundColor: theme.colors.accent,
  },
  listItemDefault: {
    backgroundColor: theme.colors.white,
  },
}));
