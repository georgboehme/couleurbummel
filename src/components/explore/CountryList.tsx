import {StackActions, useNavigation} from '@react-navigation/native';
import {ListItem, makeStyles} from '@rneui/themed';
import iso3166 from 'i18n-iso-countries';
import i18next from 'i18next';
import {memo, useContext, useMemo} from 'react';
import {FlatList} from 'react-native';
import CountryFlag from 'react-native-country-flag';

import {City, Country} from '../../types/model';
import {CountriesProps} from '../../types/screens';
import {ModelContext} from '../contexts/ModelContext';

export const CountryListScreen = ({}: CountriesProps) => <CountryList />;

/**
 * A list of countries
 */
export const CountryList = () => {
  const styles = useStyles();
  const model = useContext(ModelContext);

  const countries = useMemo(
    () =>
      model
        .allCountries()
        .sort((a, b) =>
          iso3166
            .getName(a.isoCode, i18next.language)
            .localeCompare(iso3166.getName(b.isoCode, i18next.language)),
        ),
    [model],
  );

  return (
    <>
      <FlatList
        style={styles.container}
        testID={'Screen:ExploreCountries'}
        keyExtractor={(country: Country) => country.id}
        data={countries}
        renderItem={({item}) => <CountryRow country={item} />}
      />
    </>
  );
};

const CountryRow = memo(({country}: {country: Country}) => {
  const styles = useStyles();
  const model = useContext(ModelContext);
  const navigation = useNavigation();

  if (!country.cityIds?.length) {
    return null;
  }

  return (
    <ListItem
      onPress={() => {
        const cities = country.cityIds
          ?.map(cityId => model.city(cityId))
          .filter(c => !!c)
          .map(c => c as City)
          .sort((a, b) => a.name.localeCompare(b.name));
        if (cities) {
          navigation.dispatch(
            StackActions.push('Cities', {
              cities,
            }),
          );
        }
      }}
      style={({pressed}) =>
        pressed ? [styles.listItem, styles.listItemPressed] : styles.listItem
      }>
      <CountryFlag style={styles.flag} isoCode={country.isoCode} size={25} />
      <ListItem.Content>
        <ListItem.Title style={styles.title}>
          {iso3166.getName(country.isoCode.toUpperCase(), i18next.language)}
        </ListItem.Title>
      </ListItem.Content>
      <ListItem.Chevron />
    </ListItem>
  );
});

const useStyles = makeStyles(theme => ({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  flag: {
    borderColor: theme.colors.greyOutline,
    borderWidth: 0.2,
    borderRadius: 2,
  },
  listItem: {
    borderRadius: 5,
  },
  listItemPressed: {
    backgroundColor: theme.colors.accent,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
  },
}));
