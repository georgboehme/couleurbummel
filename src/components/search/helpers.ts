import log from '../../log';
import {CouleurbummelModel} from '../../model/model';
import {Address, Colours, Corporation} from '../../types/model';

export interface ColoursQuery {
  baseColourId: string;
  colourFamilyIds: string[];
}

const coloursMatch = (
  colours: Colours,
  colourIdsToFind: string[][],
  baseColourIdsToFind?: string[],
) => {
  // If we're searching for more or fewer colours than we have, it's not a match
  if (colourIdsToFind.length !== colours.colourIds.length) {
    return false;
  }

  // If the query contains a base colour, but the corporation doesn't have one or has the wrong one, it doesn't match.
  if (
    baseColourIdsToFind?.length &&
    (!colours.baseColourId ||
      baseColourIdsToFind.indexOf(colours.baseColourId) === -1) &&
    (!colours.percussionColourId ||
      baseColourIdsToFind.indexOf(colours.percussionColourId) === -1)
  ) {
    log.debug(
      'Base colour not matching',
      colours.baseColourId,
      baseColourIdsToFind,
    );
    return false;
  }

  let index = 0;
  let coloursFound = 0;

  // See if we can find the colours in the correct order.
  for (const i in colourIdsToFind) {
    const matchingColourIds = colourIdsToFind[i];

    if (
      // Blank colours are wildcards and always match
      !matchingColourIds.length ||
      matchingColourIds.indexOf(colours.colourIds[index]) !== -1
    ) {
      coloursFound++;
    }
    index++;
  }

  return coloursFound === colourIdsToFind.length;
};

export const searchColours = (
  query: ColoursQuery,
  data: [Address, Corporation][],
  model: CouleurbummelModel,
) => {
  log.debug('Serving query', query);

  const baseColourToSearch = model.colourFamily(
    query.baseColourId,
  )?.memberColourIds;

  const coloursToSearch = (query.colourFamilyIds || []).map(
    familyId => model.colourFamily(familyId)?.memberColourIds || [],
  );

  log.debug('Resolved colours', baseColourToSearch, coloursToSearch);

  return data.filter(([_address, corporation]) => {
    return (corporation.colours || []).some(cols =>
      coloursMatch(cols, coloursToSearch, baseColourToSearch),
    );
  });
};
