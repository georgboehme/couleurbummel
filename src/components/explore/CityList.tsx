import {DrawerActions, useNavigation} from '@react-navigation/native';
import {Button, Icon, makeStyles} from '@rneui/themed';
import {ListItem} from '@rneui/themed';
import React, {memo, useContext, useMemo, useState} from 'react';

import constants from '../../constants';
import {CouleurbummelModel} from '../../model/model';
import {Address, City, Corporation} from '../../types/model';
import {CitiesScreenProps, CityListProps} from '../../types/screens';
import {
  setActiveDrawerAction,
  toggleFavouriteCityAction,
} from '../contexts/actions';
import {StateContext} from '../contexts/GlobalStateContext';
import {ModelContext} from '../contexts/ModelContext';
import fuzzySearch from '../fuzzySearch';
import {SearchableList} from '../SearchableList';
import {CorporationList} from './CorporationList';

const filterCities = (searchTerm: string, cities: City[]) =>
  cities.filter(c =>
    fuzzySearch(searchTerm.trim().toLowerCase(), c.name.toLowerCase()),
  );

export const CityListScreen = ({route}: CitiesScreenProps) => (
  <CityList cities={route.params?.cities} testID={route.params?.testID} />
);

/**
 * A list of cities
 */
export const CityList = ({cities, testID}: CityListProps) => {
  const {globalState, dispatch} = useContext(StateContext);
  const model = useContext(ModelContext);

  const allCities = useMemo(() => cities || model.allCities(), [model, cities]);

  return (
    <SearchableList
      testID={testID}
      allItems={allCities}
      filterItems={filterCities}
      showsAllResultsOnEmptySearch
      keyExtractor={(city: City) => city.id}
      renderItem={city => (
        <CityRow
          city={city}
          isFavourite={globalState.favourites.city.has(city.id)}
          onFavouritePress={async () => {
            dispatch(toggleFavouriteCityAction(city.id));
          }}
        />
      )}
      searchDelayMs={0}
    />
  );
};

/**
 * Gets an array of address and corporation tuples for a city
 *
 * @param city the city
 * @param model the model object
 */
const flatAddressesForCity = (city: City, model: CouleurbummelModel) => {
  return model
    .addressesWithCorporations(city.addressIds || [])
    .flatMap(([address, corporations]) =>
      corporations.map(c => [address, c] as [Address, Corporation]),
    ) as [Address, Corporation][];
};

const CityRow = memo(
  ({
    city,
    isFavourite,
    onFavouritePress,
  }: {
    city: City;
    isFavourite: boolean;
    onFavouritePress: () => Promise<void>;
  }) => {
    const {dispatch} = useContext(StateContext);
    const styles = useStyles();
    const model = useContext(ModelContext);
    const navigation = useNavigation();

    const [isExpanded, setIsExpanded] = useState(false);

    const addressesAndCorporations = useMemo(
      () => flatAddressesForCity(city, model),
      [city, model],
    );

    return (
      <ListItem.Accordion
        key={city.id}
        isExpanded={isExpanded}
        onPress={() => {
          setIsExpanded(!isExpanded);
        }}
        containerStyle={styles.cityContainer}
        content={
          <>
            <ListItem.Title style={styles.cityName}>{city.name}</ListItem.Title>

            <Button
              buttonStyle={styles.button}
              icon={<Icon name={constants.icons.explore.showOnMap} />}
              onPress={() => {
                const newScreenName = 'Map';
                dispatch(setActiveDrawerAction(newScreenName));
                navigation.dispatch(
                  DrawerActions.jumpTo('Map', {addressIds: city.addressIds}),
                );
                navigation.dispatch(DrawerActions.closeDrawer);
              }}
            />

            <Button
              buttonStyle={styles.button}
              icon={
                <Icon
                  name={
                    isFavourite
                      ? constants.icons.explore.favourite
                      : constants.icons.explore.nonFavourite
                  }
                />
              }
              onPress={onFavouritePress}
            />
          </>
        }>
        <CorporationList
          addressesAndCorporations={addressesAndCorporations}
          parentScreenName={'Cities'}
          searchable={false}
        />
      </ListItem.Accordion>
    );
  },
  (prevProps, nextProps) => {
    return prevProps.isFavourite === nextProps.isFavourite;
  },
);

const useStyles = makeStyles(() => ({
  cityName: {
    fontSize: 22,
    fontWeight: 'bold',
    flex: 1,
  },
  cityContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 10,
  },
  button: {
    backgroundColor: 'transparent',
  },
}));
