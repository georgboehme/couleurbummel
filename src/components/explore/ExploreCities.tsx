import {createNativeStackNavigator} from '@react-navigation/native-stack';
import React from 'react';
import {useTranslation} from 'react-i18next';

import {ScreenProps, StackScreenParamsList} from '../../types/screens';
import {CorporationDetailsScreen} from '../CorporationDetailsScreen';
import {ToggleDrawerButton} from '../navigation/ToggleDrawerButton';
import {CityListScreen} from './CityList';

const Stack = createNativeStackNavigator<StackScreenParamsList>();

/**
 * The root screen in the Cities section
 */
export const ExploreCitiesScreen = ({}: ScreenProps) => {
  const {t} = useTranslation();

  return (
    <Stack.Navigator>
      <Stack.Screen
        name="Cities"
        component={CityListScreen}
        options={{
          title: t('SCREEN_CITIES'),
          headerLeft: ToggleDrawerButton,
        }}
        initialParams={{
          testID: 'Screen:ExploreCities',
        }}
      />
      <Stack.Screen
        name="CorporationDetails"
        options={{title: t('SCREEN_DETAILS')}}
        component={CorporationDetailsScreen}
      />
    </Stack.Navigator>
  );
};
