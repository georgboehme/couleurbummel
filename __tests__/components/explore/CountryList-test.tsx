import {createNativeStackNavigator} from '@react-navigation/native-stack';
import React from 'react';

import {CityListScreen} from '../../../src/components/explore/CityList';
import {
  CountryList,
  CountryListScreen,
} from '../../../src/components/explore/CountryList';
import {StackScreenParamsList} from '../../../src/types/screens';
import {waitForPromisesToResolveWithAct} from '../../__helpers__/helpers';
import {fireEvent, render, screen, waitFor} from '../../__helpers__/test-utils';

describe('Country list', () => {
  it('renders correctly', async () => {
    render(<CountryList />);
    await waitForPromisesToResolveWithAct().then(() =>
      expect(screen.toJSON()).toMatchSnapshot(),
    );
  });

  it('pushes the cities view to the stack when a country is selected', async () => {
    const Stack = createNativeStackNavigator<StackScreenParamsList>();
    render(
      <Stack.Navigator>
        <Stack.Screen name="Countries" component={CountryListScreen} />
        <Stack.Screen name="Cities" component={CityListScreen} />
      </Stack.Navigator>,
    );

    const germanyRow = await screen.findByText('Germany');
    await waitFor(() => expect(germanyRow).toBeTruthy());

    fireEvent(germanyRow, 'press');

    const aachenRow = await screen.findByText('Aachen');
    await waitFor(() => expect(aachenRow).toBeTruthy());
  });
});
