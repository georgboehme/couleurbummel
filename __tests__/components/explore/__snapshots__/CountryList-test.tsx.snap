// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country list renders correctly 1`] = `
[
  <RCTScrollView
    data={
      [
        {
          "cityIds": [
            "city_at_wien",
          ],
          "id": "country_at",
          "isoCode": "at",
          "name": "Österreich",
          "translationKey": "COUNTRY_AT",
        },
        {
          "cityIds": [
            "city_de_aachen",
            "city_de_tübingen",
            "city_de_karlsruhe",
            "city_de_kiel",
          ],
          "id": "country_de",
          "isoCode": "de",
          "name": "Deutschland",
          "translationKey": "COUNTRY_DE",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
      }
    }
    testID="Screen:ExploreCountries"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          replaceTheme={[Function]}
          style={
            {
              "borderRadius": 5,
            }
          }
          updateTheme={[Function]}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
            testID="RNE__LISTITEM__padView"
          >
            <Image
              source={
                {
                  "testUri": "../../react-native-country-flag/dist/cjs/flags/at.png",
                }
              }
              style={
                [
                  {
                    "height": 25,
                    "width": 40,
                  },
                  {
                    "borderColor": "#bbb",
                    "borderRadius": 2,
                    "borderWidth": 0.2,
                  },
                ]
              }
            />
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <View
              replaceTheme={[Function]}
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "accent": "#3CB371",
                    "accentTransparent": "#3CB37180",
                    "background": "#ffffff",
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff3b30",
                    "fixed": {
                      "black": "#000000",
                      "darkGrey": "#393E42FF",
                      "lightGrey": "#EBEBEB",
                      "white": "#FFFFFF",
                    },
                    "grey": "#7d7d7d",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "iconOnPrimaryColourBackgroundColour": "#E1E8EE",
                    "mapMarkerCentreStroke": "#000000",
                    "mapMarkerFill": "#EBEBEB",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#008040",
                    "primaryTransparent": "rgba(0,128,64,0.75)",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "text": "#000000",
                    "warning": "#ffcc00",
                    "white": "#ffffff",
                  },
                  "mode": "light",
                  "spacing": {
                    "lg": 12,
                    "md": 8,
                    "sm": 4,
                    "xl": 24,
                    "xs": 2,
                  },
                }
              }
              updateTheme={[Function]}
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 24,
                    "fontWeight": "bold",
                  }
                }
                testID="listItemTitle"
              >
                Austria
              </Text>
            </View>
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <View
              style={
                {
                  "alignSelf": "center",
                  "overflow": "hidden",
                }
              }
              testID="RNE__ICON__CONTAINER"
            >
              <View
                accessibilityRole="button"
                android_ripple={
                  {
                    "borderless": false,
                    "color": "rgba(0, 0, 0, 0.3)",
                    "radius": -5,
                  }
                }
                disabled={false}
                replaceTheme={[Function]}
                testID="RNE__ICON__CONTAINER_ACTION"
                updateTheme={[Function]}
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                  testID="RNE__ICON"
                >
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "#D1D1D6",
                          "fontSize": 16,
                        },
                        {
                          "backgroundColor": "transparent",
                        },
                        {
                          "fontFamily": "Ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                    testID="RNE__ICON__Component"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          replaceTheme={[Function]}
          style={
            {
              "borderRadius": 5,
            }
          }
          updateTheme={[Function]}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
            testID="RNE__LISTITEM__padView"
          >
            <Image
              source={
                {
                  "testUri": "../../react-native-country-flag/dist/cjs/flags/de.png",
                }
              }
              style={
                [
                  {
                    "height": 25,
                    "width": 40,
                  },
                  {
                    "borderColor": "#bbb",
                    "borderRadius": 2,
                    "borderWidth": 0.2,
                  },
                ]
              }
            />
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <View
              replaceTheme={[Function]}
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "accent": "#3CB371",
                    "accentTransparent": "#3CB37180",
                    "background": "#ffffff",
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff3b30",
                    "fixed": {
                      "black": "#000000",
                      "darkGrey": "#393E42FF",
                      "lightGrey": "#EBEBEB",
                      "white": "#FFFFFF",
                    },
                    "grey": "#7d7d7d",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "iconOnPrimaryColourBackgroundColour": "#E1E8EE",
                    "mapMarkerCentreStroke": "#000000",
                    "mapMarkerFill": "#EBEBEB",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#008040",
                    "primaryTransparent": "rgba(0,128,64,0.75)",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "text": "#000000",
                    "warning": "#ffcc00",
                    "white": "#ffffff",
                  },
                  "mode": "light",
                  "spacing": {
                    "lg": 12,
                    "md": 8,
                    "sm": 4,
                    "xl": 24,
                    "xs": 2,
                  },
                }
              }
              updateTheme={[Function]}
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 24,
                    "fontWeight": "bold",
                  }
                }
                testID="listItemTitle"
              >
                Germany
              </Text>
            </View>
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <View
              style={
                {
                  "alignSelf": "center",
                  "overflow": "hidden",
                }
              }
              testID="RNE__ICON__CONTAINER"
            >
              <View
                accessibilityRole="button"
                android_ripple={
                  {
                    "borderless": false,
                    "color": "rgba(0, 0, 0, 0.3)",
                    "radius": -5,
                  }
                }
                disabled={false}
                replaceTheme={[Function]}
                testID="RNE__ICON__CONTAINER_ACTION"
                updateTheme={[Function]}
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                  testID="RNE__ICON"
                >
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "#D1D1D6",
                          "fontSize": 16,
                        },
                        {
                          "backgroundColor": "transparent",
                        },
                        {
                          "fontFamily": "Ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                    testID="RNE__ICON__Component"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    replaceTheme={[Function]}
    testID="Internal__Overlay"
    transparent={true}
    updateTheme={[Function]}
    visible={false}
  />,
]
`;
